### PR TITLE
runtime: merged ValidCgroupPath method

### PR DESCRIPTION
**Pull request by yaoyinnan**
Merged ValidCgroupPath method and modified unit tests.

Fixes: #8930

### CODE REVIEW

File: src/runtime/pkg/resourcecontrol/utils_linux.go [hunk 1]
```diff
@@ -21,39 +21,15 @@ import (
...
-	// In the case of an absolute path (starting with /), the runtime MUST
-	// take the path to be relative to the cgroups mount point.
-	if filepath.IsAbs(path) {
-		return filepath.Clean(path), nil
-	}
-
-	// In the case of a relative path (not starting with /), the runtime MAY
-	// interpret the path relative to a runtime-determined location in the cgroups hierarchy.
-	// clean up path and return a new path relative to DefaultResourceControllerID
-	return filepath.Join(DefaultResourceControllerID, filepath.Clean("/"+path)), nil
```
Comment (justxuewei) [inline src/runtime/pkg/resourcecontrol/utils_linux.go:44]: Why are these removed? It gives a conversion from non-systemd-cgroup path.

**Review by justxuewei — COMMENTED** [review]
Thanks @yaoyinnan! One comment here.

File: src/runtime/pkg/resourcecontrol/utils_linux.go (hunk 1 above)
Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go:44]: ValidCgroupPathV1 and ValidCgroupPathV2 are treated the same on non-systemd-cgroup paths. I didn't remove the handling of the non-systemd-cgroup path, just the duplicate part.

**Review by justxuewei — APPROVED** [review]
Lgtm, thanks @yaoyinnan!

**Review by Apokleos — COMMENTED** [review]
Thx @yaoyinnan could you please split the PR into more one commits to make it clearer ? IMO, the code related to test should be moved into one commit ?

File: src/runtime/pkg/resourcecontrol/utils_linux.go [hunk 2]
```diff
@@ -21,39 +21,15 @@ import (
 // DefaultResourceControllerID runtime-determined location in the cgroups hierarchy.
 const DefaultResourceControllerID = "/vc"
 
-// ValidCgroupPathV1 returns a valid cgroup path for cgroup v1.
+// ValidCgroupPath returns a valid cgroup path.
 // see https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#cgroups-path
-func ValidCgroupPathV1(path string, systemdCgroup bool) (string, error) {
+func ValidCgroupPath(path string, v2 bool) (string, error) {
```
Comment (justxuewei) [inline src/runtime/pkg/resourcecontrol/utils_linux.go:26]: Looks like the logic was changed due to `systemdCgroup` removal

- before: the function **throws an error** if `systemdCgroup == true` and the path is not in the format of systemd cgroup.
- now: giving a new path without any errors.

**Review by justxuewei — CHANGES_REQUESTED** [review]
Block this PR mergence until the above comment is addressed.

**Comment by justxuewei** [comment]
Btw, your email in `sign-off-by` is incorrect. You should take a look at this as well.

File: src/runtime/pkg/resourcecontrol/utils_linux.go (hunk 2 above)
Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go:26]: @justxuewei Thank you for your review, this issue is indeed worthy of discussion.

The IsSystemdCgroup() is repeated. When calling the ValidCgroupPath() method, the incoming systemdCgroup is obtained through IsSystemdCgroup(path).
This logic seems to be intended to handle the following situation: when calling the `ValidCgroupPath()`, the parameter `systemdCgroup==true` is actively passed in, but in fact, `IsSystemdCgroup(path)` is judged to be `false` inside the function.

I'm not sure if I want to keep this design from before, please give advice, thanks.

Comment (justxuewei) [reply src/runtime/pkg/resourcecontrol/utils_linux.go:26]: I still think they are different. Let us list all possible cases here.

| index | path        | systemdCgroup | result             |
|-------|-------------|---------------|--------------------|
| 1     | systemd     | true          | (origin path, nil) |
| 2     | systemd     | false         | (origin path, nil) |
| 3     | non-systemd | true          | ("", error)        |
| 4     | non-systemd | false         | (new path, nil)    |

Case 3 means I expect the path is in the format of systemd, but it isn't actually. Therefore, an error should be thrown. The case is unreachable if `systemdCgroup` is removed. Wdyt?

Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go:26]: You are right

Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go]: I fixed this question and the email issue.

**Review by justxuewei — APPROVED** [review]
Lgtm, thanks!

File: src/runtime/pkg/resourcecontrol/cgroups.go [hunk 3]
```diff
@@ -136,7 +136,8 @@ func NewResourceController(path string, resources *specs.LinuxResources) (Resour
 	var cgroupPath string
 
 	if cgroups.Mode() == cgroups.Legacy || cgroups.Mode() == cgroups.Hybrid {
-		cgroupPath, err = ValidCgroupPathV1(path, IsSystemdCgroup(path))
+		isCgroupV2 := false
+		cgroupPath, err = ValidCgroupPath(path, isCgroupV2, IsSystemdCgroup(path))
```
Comment (justxuewei) [inline src/runtime/pkg/resourcecontrol/cgroups.go:140]: Could you remove this line of code?

File: src/runtime/pkg/resourcecontrol/cgroups.go [hunk 4]
```diff
@@ -145,7 +146,8 @@ func NewResourceController(path string, resources *specs.LinuxResources) (Resour
 			return nil, err
 		}
 	} else if cgroups.Mode() == cgroups.Unified {
-		cgroupPath, err = ValidCgroupPathV2(path, IsSystemdCgroup(path))
+		isCgroupV2 := true
```
Comment (justxuewei) [inline src/runtime/pkg/resourcecontrol/cgroups.go:149]: Ditto.

File: src/runtime/pkg/resourcecontrol/cgroups.go [hunk 5]
```diff
@@ -136,7 +136,8 @@ func NewResourceController(path string, resources *specs.LinuxResources) (Resour
 	var cgroupPath string
 
 	if cgroups.Mode() == cgroups.Legacy || cgroups.Mode() == cgroups.Hybrid {
-		cgroupPath, err = ValidCgroupPathV1(path, IsSystemdCgroup(path))
+		isCgroupV2 := false
```
Comment (justxuewei) [inline src/runtime/pkg/resourcecontrol/cgroups.go:139]: Could you remove this line of code?

Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/cgroups.go]: removed

**Review by Apokleos — APPROVED** [review]
Thx @yaoyinnan LGTM

**Pull request merged**